5단계 - 로또(수동) 리뷰 요청드립니다.

yomni opened the pull request:
안녕하세요!
우선 이번주까지 개인적인 사유로 
미션 진행이 조금 더뎌져서 죄송합니다.

로또 피드백 강의 수강후 수정할 점이 보여서, 
보완 및 수동로직 추가하여 리뷰요청드립니다

5단계 미션의 주요 변경점은 inputView에서 수동입력에 대한 로또리스트를 생성하여
자동 발급 로직에 추가 보완 하였습니다.

감사합니다!

yomni committed:
docs: 5단계 기능목록 작성 및 피드백 수정사항 작성

yomni committed:
refactor: WinningLottoRank에 존재하던 View 로직을 DTO로 변경

yomni committed:
refactor: Lottos 일급컬렉션 삭제

yomni committed:
refactor: 로또 피드백 강의 부분 반영

1. 로또 : List<LottoNumber> --> Set<LottoNumber> 로 변경
2. 로또번호 : 캐싱을 Strategy보단 LottoNumber가 가지고 있도록 수정

yomni committed:
feat: 수동 구매에 대한 로직 추가

ohtaeg commented on src/main/java/lotto/LottoApplication.java:
```diff
@@ -17,30 +18,35 @@ private static void run() {
-        PurchaseAmount purchaseAmount = getPurchaseAmount(inputView, resultView);
-        Lottos lottos = getLottos(resultView, purchaseAmount);
+        PurchaseAmount purchaseAmount = getPurchaseAmount(inputView);
+        List<Lotto> manualLottos = getManualLottos(inputView, purchaseAmount);
+        List<Lotto> lottos = getLottos(resultView, manualLottos, purchaseAmount);
```
Lottos 라는 객체를 만들어주신 만큼 List\<Lotto>를 활용해주시는 것보다 Lottos로 타입을 관리해주시면 좋을 것 같아요 :)

yomni replied:
자동발급까지는 의미가없었던 Lottos 라는 타입에서  
사실 이 일급컬렉션을 없애야하나 유지해야하나 고민이 많이 됐었습니다.

이전 단계까지는 로또 시스템 내에서 Lottos란 객체에 대해 `발급한 로또들`이란 의미를 부여했었습니다.
고민한 이유는 '만약 `추가발급`이나 `발급했던 로또에 대한 삭제`라는 행위가 있어야 한다면?'  
이란 생각이 계속해서 이 Lottos를 삭제하지 못한 이유였습니다.

수동 / 자동에 대한 발급을 Lottos란 객체로 의미를 부여하여 수정해보겠습니다.

피드백 감사합니다 :)

ohtaeg commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -11,15 +9,16 @@ public class Lotto {
     public static final String EXCEPTION_MESSAGE_FOR_LOTTO_NUMBERS_DUPLICATION = "로또번호는 중복일 수 없습니다";
-    private final List<LottoNumber> lottoNumbers;
+    public static final String EXCEPTION_MESSAGE_FOR_LOTTO_NUMBERS_EMPTY = "로또번호를 입력해주세요.";
+    private final Set<LottoNumber> lottoNumbers;
```
Set으로 바꿔주셨군요 👍

ohtaeg commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -31,10 +30,16 @@ public boolean getMatchBonus(LottoNumber bonusNumber) {
         validateLottoNumbersSize(lottoNumbers);
     }
 
+    private static void validateNullPointException(List<LottoNumber> lottoNumbers) {
+        Optional.ofNullable(lottoNumbers.size())
+                .orElseThrow(() -> new IllegalArgumentException(EXCEPTION_MESSAGE_FOR_LOTTO_NUMBERS_EMPTY));
+    }
```
NPE validation을 해주셨는데 실제로 null을 파라미터로 넣게되면 NPE가 발생하는 것 같아요..!
한번 확인해주시면 감사하겠습니다 :)

<img width="704" alt="image" src="https://user-images.githubusercontent.com/20432323/200311291-6f9a104f-ca21-4c12-8e84-959beef4285f.png">

yomni replied:
앗! 이부분은 제가 리팩토링 하면서 테스트를 누락한 것 같습니다.
TDD 미션 답게 테스트 코드부터 추가해보겠습니다.

ohtaeg commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -1,14 +1,22 @@
     public static final int MINIMUM_LOTTO_NUMBER = 1;
     public static final int MAXIMUM_LOTTO_NUMBER = 45;
     public static final String EXCEPTION_MESSAGE_FOR_OUT_OF_LOTTO_NUMBER_BOUNDARY = "로또번호는 " + MINIMUM_LOTTO_NUMBER + "과 " + MAXIMUM_LOTTO_NUMBER + "사이의 숫자이어야 합니다.";
+    private static final Map<Integer, LottoNumber> lottoNumberMap = IntStream.rangeClosed(MINIMUM_LOTTO_NUMBER, MAXIMUM_LOTTO_NUMBER)
+            .mapToObj(LottoNumber::new)
+            .collect(Collectors.toMap(LottoNumber::lottoNumber, e -> e));
```
캐싱을 LottoNumber에서 관리하도록 👍

ohtaeg commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -1,6 +1,7 @@
+import java.util.List;
 import java.util.Map;
 
 public class WinningResult {
```
로또 구매시 1등이 여러번 되어서 때부자를 예상했는데 아쉽게도 적자가 나버렸습니다 😭 
<img width="306" alt="image" src="https://user-images.githubusercontent.com/20432323/200312030-26d94968-f50a-4824-b408-63c84acde97f.png">

yomni replied:
primitive 타입의 오버플로우를 고려 안했네요 😭
오버플로우에 대한 보완로직도 추가하겠습니다! 감사합니다!

ohtaeg commented on src/test/java/lotto/domain/LottoSellerTest.java:
```diff
@@ -15,16 +18,16 @@ public class LottoSellerTest {
     @ValueSource(ints = {-1000, 900, 800, 500})
     void 로또판매_금액부족예외(int purchaseAmount) {
-        assertThatThrownBy(() -> LottoSeller.sellLottos(new PurchaseAmount(purchaseAmount)))
+        assertThatThrownBy(() -> LottoSeller.sellLottos(new PurchaseAmount(purchaseAmount), new ArrayList<>()))
```
현재 빈 리스트를 넣고 테스트를 해주고 계시는데요!
현재 방법도 좋지만 LottoSeller에서 자동 판매와 수동 판매 메서드를 분리해준다면
반환 값은 같을지라도 확실하게 메서드 명과 파라미터를 일치시킬 수 있을 것 같아요 :)

물론 별도로 수동구매로 산 로또와 자동 구매로 산 로또를 합쳐주는 로직을 구현해주셔야하는 단점이 있네요..!
용민님은 이에 대해 어떻게 생각하시나요 :) ?

yomni replied:
TDD 강의에서 리팩토링은 점진적으로 확장하는 방안으로 고려해야한다는 것이 떠오르네요.

말씀하신대로 `자동 판매`만 최초 구현했던 로직에 대하여 `수동 판매`도 고려해야하는 요구사항이 추가로 생겼다면,
리팩토링의 관점에서 메소드를 분리하여, 확장하는 것이 좋을 것 같습니다!

좋은 피드백 감사합니다! 🧐

ohtaeg commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -23,6 +31,15 @@ public int lottoNumber() {
     }
 
+    public static Map<Integer, LottoNumber> lottoNumberMap() {
+        return new HashMap<>(lottoNumberMap);
```
Collections.unmodifiableMap()을 활용해보시는 것도 좋을 것 같아요 :)

ohtaeg reviewed (changes requested):
5단계도 잘해주셨어요! 👍 
소소한 코멘트들을 드렸는데 확인해주시면 감사하겠습니다 :)
마지막 피드백이 될 것 같아요! 마지막까지 화이팅입니다 🔥

yomni committed:
docs: 5단계 피드백에 대한 추가 기능목록 작성

yomni committed:
refactor: 일급컬렉션 객체의 copy를 Collections.unmodifiable*() 메소드로 수정

yomni committed:
refactor: Lotto NPE에 대한 테스트코드 및 로직 보완

yomni committed:
refactor: List<Lotto>를 필드로 가지는 일급컬렉션으로 Lottos 재정의

yomni committed:
refactor: `자동판매`와 `수동판매` 로직 분리, 통합 로직 추가

yomni committed:
fix: 수익률 집계 시 overflow 수정

yomni committed:
refactor: 로또 피드백 강의에 있었던 Lottos의 match 로직을 제가 WinningResult와 Lottos에 맞게 반영하였습니다.
추가로, View로의 역할을 모두 이전하여 WinningResultDTO 객체를 삭제하였습니다.
마지막으로, WinningResult의 report* 메소드 로직을 정리하여 매개변수 갯수를 극한까지 줄였습니다.

yomni committed:
fix: 자동발급 갯수 입력값 오류 부분 수정

yomni committed:
style: text align 및 import 패키지 정리

yomni commented:
피드백 주신 부분과 더불어 로또 피드백 강의를 다시 정독하며,
기능 목록을 다시 정의하였습니다 :)

1. Application.run() view의 메소드들의 매개변수를 줄일 수 있는 방향으로 보완하였습니다.
2. Lottos 타입을 추가하였습니다.  
    - 특히 이전 소스까지는 단순히 Wrapping 했던 Lottos 객체에서의 WinningResult를 반환하는 match(WinningLotto)라는 행위를 가지고 있도록 리팩토링 했습니다.
3. WinningResult의 수익률 계산시 int --> long 으로 타입 변경하여, overflow에 대하여 보완로직을 적용했습니다.
4. LottoSeller의 수동, 자동 판매를 분리하였고, 통합 메소드도 정의하였습니다.
5. 일급컬렉션의 반환 메소드에 대하여 가이드 주신 `Collections.unmodifiable*()`메소드로 적용하였습니다.

ohtaeg commented on src/main/java/lotto/domain/LottoSeller.java:
```diff
@@ -5,17 +5,23 @@
-        List<Lotto> lottos = new ArrayList<>(manualLottos);
-        for (int i = 0; i < lottoTicketCount - manualLottoTicketCount; i++) {
-            lottos.add(getLotto(new AutoLottoCreateStrategy()));
+    public static Lottos sellAutoLottos(int lottoTicketCount) {
+        List<Lotto> autoLottos = new ArrayList<>();
+        for (int i = 0; i < lottoTicketCount; i++) {
+            autoLottos.add(getLotto(new AutoLottoCreateStrategy()));
```
자동 생성 전략을 통해 자동 판매 👍 
소소하지만 AutoLottoCreateStrategy과 같은 자동 판매 전략을 인터페이스로 추출해서 외부로부터 주입받도록 한다면
로또 판매자는 좀 더 다양하게 로또를 생성할 수 있지 않을까 합니다 :)
가령 현실 세계에서의 로또 기계를 예로 들어보면 짝수 위주, 홀수 위주, 10번대 위주로의 로또 자동 생성 선택지가 있어서요 :)

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Lottos {
```
애매했던 Lottos가 드디어 제 역할을 하게 되었군요 😄

ohtaeg commented on src/main/java/lotto/domain/WinningLottoRank.java:
```diff
@@ -12,9 +12,9 @@ public enum WinningLottoRank {
     public static final int MATCH_COUNT_FOR_SECOND_OR_THIRD = 5;
     private final int matchCount;
-    private final int reward;
+    private final long reward;
```
👍

ohtaeg commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -16,22 +18,29 @@ public WinningResult() {
+
+    private long totalReward() {
+        return ranks.keySet()
+                .stream()
+                .mapToLong(rank -> rank.getReward() * ranks.get(rank))
+                .sum();
+    }
+
+    private int totalPurchaseAmount() {
+        return ranks.keySet()
+                .stream()
```
stream 활용 👍

ohtaeg reviewed (approved):
마지막까지 피드백 반영을 잘해주셨어요 👍 
로또 미션 완주를 축하드립니다 🎉 
리뷰를 잘 드렸는지 모르겠네요 ㅎㅎ; 
드렸던 피드백들이 조금이나마 용민님께 보탬이 되었으면 하네요..⭐

끝까지 제 피드백을 따라와주셔서 감사드리며, 함께하여 영광이였습니다!!
다음 미션들도 응원하겠습니다 🥂
고생하셨습니다 🙇
